trpcultivate_phenotypes/tests/src/Kernel/Validators/MetadataInputTest.php

```diff
@@ -0,0 +1,180 @@
+
+    // Install module configuration.
+    $this->installConfig(['trpcultivate_phenotypes']);
+
+    // Test Chado database.
+    // Create a test chado instance and then set it in the container for use by our service.
+    $this->connection = $this->createTestSchema(ChadoTestKernelBase::PREPARE_TEST_CHADO);
+    $this->container->set('tripal_chado.database', $this->connection);
+
+    // Set plugin manager service.
+    $this->plugin_manager = \Drupal::service('plugin.manager.trpcultivate_validator');
+
+    $genus = 'Tripalus';
+    // Create our organism and configure it.
+    $organism_id = $this->connection->insert('1:organism')
+      ->fields([
+        'genus' => $genus,
+        'species' => 'databasica',
+      ])
+      ->execute();
+
+    $this->assertIsNumeric($organism_id, 'We were not able to create an organism for testing (configured).');
+    $this->test_genus['configured'] = $genus;
+    $this->setOntologyConfig($this->test_genus['configured']);
+
+    // Create another organism but create a configuration item
+    // where the cv for trait is set to 0 - not configured.
+    $genus = 'notconfiggenus';
+    $organism_id = $this->connection->insert('1:organism')
+      ->fields([
+        'genus' => $genus,
+        'species' => 'databasica',
+      ])
+    ->execute();
+
+    $this->assertIsNumeric($organism_id, 'We were not able to create an organism for testing (not configured).');
+    $this->test_genus['not-configured'] = $genus;
+    $config = \Drupal::configFactory()->getEditable('trpcultivate_phenotypes.settings')
+      ->set('trpcultivate.phenotypes.ontology.cvdbon.' . $this->test_genus['not-configured'], ['trait' => 0]);
+    
+    // Set terms configuration.
+    $this->setTermConfig();
+  }
+
+  /**
+   * Test genus input.
+   */
+  public function testGenusInput() {
+    // Create a plugin instance for this validator
+    $validator_id = 'trpcultivate_phenotypes_validator_genus_exists';
+    $instance = $this->plugin_manager->createInstance($validator_id);
+    
+    // The validator expects the form values entered into the form elements
+    // through the $form_state.
+    
+    // Create form fields and set value.
+    // This would have been a form that has been submitted.
+    $form_state = new FormState();
+    // A random field.
+    $form_state->setValues(['project' => uniqid()]);
+
+    // Test items that will throw exception:
+    // 1. Passing object or the entire $form_state.
+    // 2. Failed to implement a form field element with genus name/key.
+
+    // Test passing the $form_state.
+    $exception_caught  = FALSE;        
+    try {
+      $instance->validateMetadata($form_state);
+    }
+    catch (\Exception $e) {
+      $exception_caught  = TRUE;
+    }
+    
+    $this->assertTrue($exception_caught, 'Failed to catch exception when passing a $form_state to genus metadata validator.');
+    $this->assertStringContainsString('Unexpected object type was passed', $e->getMessage(), 
```

Comment:
Using $e here is out of scope. Instead you should have the exception_caught and an exception_message that are set in the catch clause.

Reply:
Yikes... strange it did not complain. This is out of scope too https://github.com/TripalCultivate/TripalCultivate-Phenotypes/blob/d35be827579f35939be3f753aa1ab11fe79d5b5b/trpcultivate_phenotypes/tests/src/Kernel/Validators/ValidatorTraitImporterTest.php#L151

Reply:
Yes, I found this myself, PHP is very lax on scope so it will let you use it. However, if this test ever fails and an exception isn't thrown then you don't get the nice assert but rather get a variable is not set error 🤪  I know from experience 🙈